[ARM64] Use link register instead of pinning a register for materializing big load constants

hyjorc1 opened the pull request:
#### fb402aa86c16383969af124c016231a45898ec32
<pre>
[ARM64] Use link register instead of pinning a register for materializing big load constants
<a href="https://bugs.webkit.org/show_bug.cgi?id=228710">https://bugs.webkit.org/show_bug.cgi?id=228710</a>

Reviewed by Saam Barati.

Previously, we pin a register as a temp for materializing a large constant that cannot fit in
Load/Store imm form. This is not efficient since the register allocator has one less register
to allocate from. To solve this problem, we should switch to using the link register as the temp
on ARM64.

* Source/JavaScriptCore/b3/B3Common.cpp:
(JSC::B3::linkRegister):
(JSC::B3::pinnedExtendedOffsetAddrRegister): Deleted.
* Source/JavaScriptCore/b3/B3Common.h:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::callFrameAddr):
* Source/JavaScriptCore/b3/air/AirCode.cpp:
(JSC::B3::Air::Code::Code):
* Source/JavaScriptCore/b3/air/AirLowerStackArgs.cpp:
(JSC::B3::Air::lowerStackArgs):
* Source/JavaScriptCore/ftl/FTLThunks.cpp:
(JSC::FTL::genericGenerationThunkGenerator):

Canonical link: <a href="https://commits.webkit.org/253105@main">https://commits.webkit.org/253105@main</a>
</pre>

justinmichaud commented on Source/JavaScriptCore/b3/B3Common.cpp:
```diff
@@ -70,10 +70,10 @@ bool shouldSaveIRBeforePhase()
 }
 
-std::optional<GPRReg> pinnedExtendedOffsetAddrRegister()
+std::optional<GPRReg> linkRegister()
```
Do we want some kind of RAII object to make sure two places don't do this with overlapping live-ranges?

saambarati commented on Source/JavaScriptCore/b3/B3Common.cpp:
```diff
@@ -70,10 +70,10 @@ bool shouldSaveIRBeforePhase()
 }
 
-std::optional<GPRReg> pinnedExtendedOffsetAddrRegister()
+std::optional<GPRReg> linkRegister()
```
I would keep mostly the same as the old name, but no need to say `pinned` anymore. Maybe just `extendedOffsetAddrRegister`

hyjorc1 replied:
updated

saambarati commented on Source/JavaScriptCore/b3/B3Common.cpp:
```diff
@@ -70,10 +70,10 @@ bool shouldSaveIRBeforePhase()
+    return MacroAssembler::linkRegister;
 #elif CPU(X86_64)
     return std::nullopt;
 #else
```
Can we just ASSERT that this is only ever called on arm64 now? Also, if we do that, no need to return optional. Just return a GPRReg now, since on arm64 it'll be correct. And we'll crash on non arm64

hyjorc1 replied:
updated

saambarati reviewed (approved):
r=me with comments

hyjorc1 committed:
[ARM64] Use link register instead of pinning a register for materializing big load constants
https://bugs.webkit.org/show_bug.cgi?id=228710

Reviewed by Saam Barati.

Previously, we pin a register as a temp for materializing a large constant that cannot fit in
Load/Store imm form. This is not efficient since the register allocator has one less register
to allocate from. To solve this problem, we should switch to using the link register as the temp
on ARM64.

* Source/JavaScriptCore/b3/B3Common.cpp:
(JSC::B3::linkRegister):
(JSC::B3::pinnedExtendedOffsetAddrRegister): Deleted.
* Source/JavaScriptCore/b3/B3Common.h:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::callFrameAddr):
* Source/JavaScriptCore/b3/air/AirCode.cpp:
(JSC::B3::Air::Code::Code):
* Source/JavaScriptCore/b3/air/AirLowerStackArgs.cpp:
(JSC::B3::Air::lowerStackArgs):
* Source/JavaScriptCore/ftl/FTLThunks.cpp:
(JSC::FTL::genericGenerationThunkGenerator):

Canonical link: https://commits.webkit.org/253105@main
